File: src/domain/Judge.ts
import type { Board } from './Board'
import type { Direction, Player, Position, WinningLine } from './types'

const DIRECTIONS: { dr: number; dc: number; dir: Direction }[] = [
  { dr: 0, dc: 1, dir: 'Horizontal' },
  { dr: 1, dc: 0, dir: 'Vertical' },
  { dr: 1, dc: 1, dir: 'LeftDiagonal' },
  { dr: 1, dc: -1, dir: 'RightDiagonal' }
]

function playerToCell(player: Player): 'Black' | 'White' {
  return player === 'Black' ? 'Black' : 'White'
}

function countDirection(
  board: Board,
  pos: Position,
  player: Player,
  dr: number,
  dc: number
): number {
  const target = playerToCell(player)
  let count = 0
  let curRow = pos.row + dr
  let curCol = pos.col + dc
  while (board.isValidPosition({ row: curRow, col: curCol }) && board.getCell({ row: curRow, col: curCol }) === target) {
    count++
    curRow += dr
    curCol += dc
  }
  return count
}

export function checkWin(
  board: Board,
  lastMove: Position,
  player: Player
): WinningLine | null {
  for (const { dr, dc, dir } of DIRECTIONS) {
    const forward = countDirection(board, lastMove, player, dr, dc)
    const backward = countDirection(board, lastMove, player, -dr, -dc)
    const total = 1 + forward + backward

    if (total >= 5) {
      return {
        direction: dir,
        start: {
          row: lastMove.row - dr * backward,
          col: lastMove.col - dc * backward
        },
        length: total,
        player
      }
    }
  }
  return null
}

export function checkDraw(board: Board): boolean {
  return board.isFull()
}

File: src/Judge.cpp
#include "Judge.h"

CellState Judge::playerToCell(Player player) {
    return (player == Player::Black) ? CellState::Black : CellState::White;
}

int Judge::countDirection(const Board& board, Position pos, Player player, int dr, int dc) {
    CellState target = playerToCell(player);
    int count = 0;
    Position cur{pos.row + dr, pos.col + dc};
    while (board.isValidPosition(cur) && board.getCell(cur) == target) {
        ++count;
        cur.row += dr;
        cur.col += dc;
    }
    return count;
}

std::optional<WinningLine> Judge::checkWin(const Board& board, Position lastMove, Player player) {
    const int dirs[4][2] = {
        {0, 1},
        {1, 0},
        {1, 1},
        {1, -1}
    };
    const Direction dirEnums[4] = {
        Direction::Horizontal,
        Direction::Vertical,
        Direction::LeftDiagonal,
        Direction::RightDiagonal
    };

    for (int i = 0; i < 4; ++i) {
        int dr = dirs[i][0];
        int dc = dirs[i][1];
        int forward = countDirection(board, lastMove, player, dr, dc);
        int backward = countDirection(board, lastMove, player, -dr, -dc);
        int total = 1 + forward + backward;

        if (total >= 5) {
            Position start{
                lastMove.row - dr * backward,
                lastMove.col - dc * backward
            };
            return WinningLine{dirEnums[i], start, total, player};
        }
    }
    return std::nullopt;
}

bool Judge::checkDraw(const Board& board) {
    return board.isFull();
}